src/VecSim/vec_sim_index.h

```diff
@@ -55,4 +57,37 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
     // Static class functions
     static timeoutCallbackFunction timeoutCallback;
     static void setTimeoutCallbackFunction(timeoutCallbackFunction callback);
+
+    template <typename DataType>
+    static void NormalizeVector(const void *input_vector, size_t dim, DataType *output = NULL);
 };
+
+static const double MAX_FP64 = std::numeric_limits<double>::max();
+
+template <typename DataType>
+void VecSimIndexAbstract::NormalizeVector(const void *input_vector, size_t dim, DataType *output) {
+
+    // if output is not NULL and doesn't point to the same memory area as input.
+    if (output && output != (DataType *)input_vector) {
+        memcpy(output, input_vector, dim * sizeof(DataType));
+    } else { // else we are normalizing inplace.
+        output = (DataType *)input_vector;
```

Comment:
You are breaking `const` guard here. If you need to copy, copy outside of the function. recall the function purpose is to normalize.